compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt

```diff
@@ -0,0 +1,124 @@
+        val enumMemberProperties = enumClass.declarations.filterIsInstance<FirProperty>()
+        val enumEntries = enumClass.declarations.filterIsInstance<FirEnumEntry>()
+
+        // Members inside the companion object of an enum class
+        if (containingClass == enumClass.companionObject) {
+            // Uninitialized from the point of view of members or enum entries of that enum class
+            if (accessedContext in enumMemberProperties || accessedContext.correspondingEnumEntry(enumEntries) != null) {
+                if (referencedDeclaration is FirProperty) {
+                    // From KT-11769
+                    // enum class Fruit(...) {
+                    //   APPLE(...);
+                    //   companion object {
+                    //     val common = ...
+                    //   }
+                    //   val score = ... <!>common<!>
+                    // }
+                    reporter.reportOn(source, FirErrors.UNINITIALIZED_VARIABLE, referencedDeclaration.symbol, context)
+                } else {
+                    // enum class EnumCompanion2(...) {
+                    //   INSTANCE(<!>foo<!>())
+                    //   companion object {
+                    //     fun foo = ...
+                    //   }
+                    // }
+                    // <!>Companion<!>.foo() v.s. <!>foo<!>()
+                    if ((expression.explicitReceiver as? FirResolvedQualifier)?.symbol?.fir == enumClass.companionObject) {
+                        reporter.reportOn(
+                            expression.explicitReceiver!!.source,
+                            FirErrors.UNINITIALIZED_ENUM_COMPANION,
+                            enumClass.symbol,
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_COMPANION, enumClass.symbol, context)
+                    }
+                }
+            }
+
+        }
+
+        // The enum entries of an enum class
+        if (referencedDeclaration in enumEntries) {
+            val callee = referencedDeclaration as FirEnumEntry
+            // Uninitialized from the point of view of members of that enum class
+            if (accessedContext in enumMemberProperties) {
+                // From KT-6054
+                // enum class MyEnum {
+                //   A, B;
+                //   val x = when(this) {
+                //     <!>A<!> -> ...
+                //     <!>B<!> -> ...
+                //   }
+                // }
+                reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_ENTRY, callee.symbol, context)
+            }
+            // enum class D(...) {
+            //   D1(<!>D2<!>),
+            //   D2(...)
+            // }
+            accessedContext.correspondingEnumEntry(enumEntries)?.let { caller ->
```

Comment:
If you have `correspondingEnumEntry` return something like an `indexOf` (-1 if not in `enumEntries`), then in here you can just loop backwards from the index of the caller until you find the callee (or not). Or you could add some comments below; it was confusing to me.

Reply:
Hm, looping backwards sounds more confusing to me, e.g., if the other one does appear in front of the starting point. Added comments about how `indexOf` can be used here, as well as why I chose this approach.

Basically, for the given two elements of interest, iterating the list and finding any one that precedes. This way, we can iterate just once, up to that first appearance.